jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientURITest.java

```diff
@@ -63,6 +67,51 @@ public void testIPv6Host() throws Exception
         Assert.assertEquals(HttpStatus.OK_200, request.send().getStatus());
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void testIDNHost() throws Exception
+    {
+        startClient();
+        client.newRequest("http://пример.рф"); // example.com-like host in IDN domain
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testIDNRedirect() throws Exception
+    {
+        String exampleHost = "http://пример.рф"; // example.com-like host in IDN domain
+        String incorrectlyDecoded = new String(exampleHost.getBytes(StandardCharsets.UTF_8), StandardCharsets.ISO_8859_1);
+
+        IncorrectRedirectServer server = new IncorrectRedirectServer(exampleHost);
+        startClient();
+        server.start();
+
+        try
+        {
+            ContentResponse response = client.newRequest("localhost", server.localPort())
+                    .timeout(5, TimeUnit.SECONDS)
+                    .followRedirects(false)
+                    .send();
+
+            HttpField location = response.getHeaders().getField(HttpHeader.LOCATION);
+            Assert.assertEquals(incorrectlyDecoded, location.getValue());
```

Comment:
This first request must be outside the `try` block, otherwise if an `ExecutionException(IllegalArgumentException)` is thrown for some other reason the test wrongly passes.